xml/vk.xml

```diff
@@ -4861,6 +4861,12 @@ typedef void <name>CAMetalLayer</name>;
             <member><type>VkBool32</type>                           <name>formatA4R4G4B4</name></member>
             <member><type>VkBool32</type>                           <name>formatA4B4G4R4</name></member>
         </type>
+        <type category="struct" name="VkPhysicalDeviceDrmPropertiesEXT" returnedonly="true" structextends="VkPhysicalDeviceProperties2">
+            <member values="VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_DRM_PROPERTIES_EXT"><type>VkStructureType</type> <name>sType</name></member>
+            <member><type>void</type>*                              <name>pNext</name></member>
+            <member><type>int32_t</type>                           <name>primary_minor</name></member>
+            <member><type>int32_t</type>                           <name>render_minor</name></member>
```

Comment:
major/minor numbers are defined as unsigned :+1: 
```
unsigned int major(dev_t dev);
unsigned int minor(dev_t dev);
```

Are you planning on using `int32_t` as a way to report error/unknown?

Reply:
Yeah, my plan was to make -1 mean "there is no node of this type".

Reply:
In practice, aren't major/minor <= 255?  I had thought using int32_t and -1 for unknown would be a good idea personally, but there could always be a separate explicit "not a DRM device" (or separate "no render node" and "no primary node" if appropriate) field as well if others prefer.

The MR looks like a good start to me.  Will review again when the spec language is filled in.

Reply:
>In practice, aren't major/minor <= 255?

On [Linux](https://sourceware.org/git/?p=glibc.git;a=blob;f=bits/typesizes.h;h=8f16903a216aae5d0fb58b2aae4f00d6972e9b00;hb=HEAD#l29) and [FreeBSD](https://github.com/freebsd/freebsd/blob/4bfbbd74a8206fc222ca8b3ee874a963ee6900e0/sys/sys/_types.h#L113) user-space, `dev_t` is a 64-bit type. However [in the Linux kernel](https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/kdev_t.h) major/minor can only be <= 255 as you said. On FreeBSD [the full 64-bit value may be used](https://github.com/freebsd/freebsd/blob/b3866f4bf2a28d20aa9a91a6f8737394e3546837/sys/sys/types.h#L381). OpenBSD seems to have a minor that can [go over 255](https://github.com/openbsd/src/blob/13001500586b3e292b0572e147c78146228ce70c/sys/sys/types.h#L211).

On Linux and OpenBSD, major/minor are defined to return an `unsigned int`. On FreeBSD, they are defined to return an `int`.

Ultimately, major/minor are system-specific and may be (1) bigger than `int32_t` (2) undefined (libdrm doesn't (yet) require that `dev_t` can be split into major/minor). One idea would be to embed `dev_t` fields in the struct directly, but I'm not sure it's a good idea (and this wouldn't make it any easier to transfer over the wire, e.g. for Wayland).

Overall, I'm starting to re-consider the use of major/minor, because finding a way to use them that doesn't put us into a corner is tricky and it sounds like more pain with little benefit. Maybe we should do like the EGL extension and provide a path as a string. The benefits of major/minor over a device node path I can think of are:

- Not tied to a specific file-system layout, so works when exchanging them between namespaces that may have `/dev/dri` set up differently. But we don't really care about this for Vulkan, since the Vulkan driver and the app run in the same process. Also a lot of clients expect `/dev/dri` to be set up correctly anyways…
- Can easily compare two devices for equality without having to deal with symlinks/etc. Side note, major/minor are still different for the primary and render nodes. If the app gets a path, it needs to open it, use `drmGetDevice` and then `drmDevicesEqual` to compare them. Or it could just use `realpath`.

I'm also not 100% clear of how we would "return" a string from `vkGetPhysicalDeviceProperties2`. Use a `const char *` field?

Reply:
> I'm also not 100% clear of how we would "return" a string from vkGetPhysicalDeviceProperties2. Use a const char * field?

We typically do this by populating a fixed size array. For example, VkPhysicalDeviceDriverProperties in VK_KHR_driver_properties does this:

```
// Provided by VK_VERSION_1_2
typedef struct VkPhysicalDeviceDriverProperties {
    VkStructureType         sType;
    void*                   pNext;
    VkDriverId              driverID;
    char                    driverName[VK_MAX_DRIVER_NAME_SIZE];
    char                    driverInfo[VK_MAX_DRIVER_INFO_SIZE];
    VkConformanceVersion    conformanceVersion;
} VkPhysicalDeviceDriverProperties;
```

Reply:
POSIX seems to have failed us here.

I don't have quite as bleak an outlook though. On both Linux and the BSDs, major and minor are values that fit in 32-bits, and likely always will be unless those OS's want to burn down the world. I'd suggest we simply return int64_t major/minor numbers, assert their range is limited such that the value can always be safely cast to the type used by the major() and minor() functions on the OS on which the Vulkan instance is running, and use a separate boolean field to indicate cases where there is no corresponding device node.

I think returning a string would be mildly problematic because:

* It may be hard to pick a fixed length for a path string. I think POSIX requires things be under /dev, and I'd guess most OS's limit themselves to PATH_MAX as a practical matter, but there's no real guarantee AFAIK.
* I'm assuming there may be cases where sandboxes disallow access to /dev or even present a very lightly populated copy of /dev to apps, but apps may still want to match some piped-in FDs against the Vulkan device using fstat().
* In practice, our Linux driver (and likely others on Linux) has to make assumptions at build time about the layout of /dev or add a dependency on udev and/or libdrm to provide such a string, and this sort of tradeoff/choice is better left to apps IMHO.

Reply:
OK, all of that makes sense to me. major/minor isn't the best solution, but yeah the alternatives aren't great either.

Just to make sure we haven't missed an opportunity: can you confirm it'd be cumbersome to embed some `dev_t` fields in the Vulkan struct? This would require to `#include <sys/types.h>` on Unix, but would avoid using any system-specific API ([`dev_t` is POSIX](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/types.h.html)).